monitoring/prometheus/alerts/ceph_default_alerts.yml

```diff
@@ -214,6 +214,17 @@ groups:
             will be full in less than 5 days assuming the average fill-up
             rate of the past 48 hours.
 
+      - alert: Different MTU Size
+        expr: node_network_mtu_bytes != ignoring(instance,interface,job) group_left() quantile(0.5, node_network_mtu_bytes)
```

Comment:
A few things that caught my eye.

The `interface` label of the `node_network_mtu_bytes` metric is only called `interface` in node-exporter version 0.17. Since v0.18 it is called `device`. Hence, this alert won't work with any node-exporter version starting at v0.18 (I've tested it up to 1.01).

Ignoring every label maybe easier and possibly safer with `on()` instead of `ignoring(instance,interface,job)`.

On my machine, the Prometheus expression for the alert always returns the `lo` device with value `65536`, so it would always trigger this alert, which it probably shouldn't. Maybe `lo` should be ignored.

**quantile**

I assume the `quantile` function would work well in most cases. However, in some cases it might not work as expected, resulting in hard to understand behavior.

For one, it does not return the setting that is most frequently used. Stating that in the alerts message might then be confusing when it actually isn't the case. I assume the values I use here are very unlikely to be met in the wild, but just for the purpose of demonstration, I'll use those.
Assume these fictional values for calculating the quantile: `[1000, 1000, 1500, 2000, 2000]`. The result would be `1500` although is occurs only once.

Please also consider that the 50% quantile is calculated as the average of the two items in the middle of the sorted input array when the length of the array is even. In the case of a user being unlucky and having two configurations evenly distributed among all machines, the quantile will return a value that is not equal to any other value, resulting in alerts for all devices being fired. For example `[1000, 1000, 1500, 1500]` will inevitably return `1250` which won't match any configuration of any network device.

Nit: It would be helpful to be told which value is actually being used by the majority of devices in the message of the alert, which, I think, is something that the `quantile` function cannot calculate. On the other hand, the 50% quantile is the median of all values. If we do stick with `quantile` for this alert, it might still be helpful to tell the user which median has been calculated and used as the basis for the comparison of this alert.

For a proper solution I wonder what the alert should allow/disallow. Should every single deviation from the standard be fired as alert? Or should there be an allowed margin of error?

I wonder if there's an option better than using `quantile` for this problem.

Reply:
Thanks @p-se for this detailed review. As per your observations, I will try to find a better solution for this.

Reply:
@p-se: the `quantile(0.5, ...)` here was to get the median of the MTUs as we are trying to detect outliers: that is, deployments where most interfaces are set to a given MTU size (mostly jumbo frames), but due to day 2 ops or other reasons, some of them may get reset to the default 1500.

So that list would rather be something like:
- `[1500, 1500, 1500, 1500, 9000, 1500, 9000, 1500]`: Regular-frame deployment: in this case the anomaly is 9000.
- `[9000, 9000, 1500, 9000, 9000, 1500, 9000, 9000]`: Jumbo-frame deployment: in this case the anomaly is 1500.

Reply:
And we may lack a better way to do it. So even if the median will not always be correct, at least theoretically and maybe eventually once or twice in reality, too, it might be the best choice. I understand that and would accept that, but still wouldn't be 100% satisfied with that solution. 

I've also thought about different options but haven't found a better approach so far.

Reply:
If we stick with `quantile` I'd opt for adapting the alert's message to not say `majority` but `median` then. Although values could be used in the message which we do not want, the message itself would probably always be true when referring to the median.

Reply:
@p-se even if the median is not correct for vectors with an even number of samples and a 50% setting for each, the alert should still be triggered, since we are talking about scenarios like these (we cannot infer which setting is correct though):
`[1500, 9000]`
`[1500, 1500, 9000, 9000]`

But it won't be triggered when:
`[1500, 9000, 9000, 9000]`
`[1500, 1500, 9000, 9000]`

Reply:
Expressions that will end up doing something like `value != 5250` when there's no value like `5250` will fire multiple times. Alerts will be fired for values which should've been considered correct, at least for one half or the other (half). That means, if the median is calculated in a way that it _doesn't match any existing value_, alerts will be fired for **all** values, even those which are correct. This is the case in the first example.

The problem with the second example is obvious. We've an alert that won't work in some (though rare) cases and it, as I said in my first response, might not be obvious why it doesn't. That troubles me. 

How should we respond to a user/customer complaining about this alert which
1. has not fired although the user expected it to or
2. has fired alerts for basically all values, although half of it should've been considered correct?

In both cases we couldn't offer a solution (using the implementation with the median). That's why I'd prefer to have the alert

1. fire in all cases where it is supposed to do that **and**
2. never fire alerts for **all** configured MTU values **or maybe**
3. we simply shouldn't introduce an alert which is hard to understand and won't work in some cases.

That leaves me with the question:

**Do you think we should have an alert that works correctly _most of the time_, but _sometimes_ it doesn't, because it doesn't fire alerts or it fires too many, without the option to fix that?**

And by not being able to fix that I mean, that if we decide to accept this solution because we know that there's no better, we'll end up in that situation. The only way to prevent that is to find another solution (now) or prevent this one from being merged.

Is this solution good enough for you to be merged?

Reply:
@p-se the only (realistic) situation where this is going to pop up is when exactly 50% of the nodes have a given setting and the other half have the another one. In such uncommon situation it is 'statistically' impossible to discern which MTU setting is correct, unless we assume that 9000 bytes is preferred (and that implies hardcoding a 9000 somewhere). The only troublesome scenario is when the cluster was just deployed and growing from 1 node to 2, or from 3 to 4, where there are chances to have an equal number of 1500 and 9k.

Alternatively, instead of the median/.5-quantile we could use the mode: `topk(1, vector)`?

Reply:
> Alternatively, instead of the median/.5-quantile we could use the mode: `topk(1, vector)`?

That might work, though I haven't created a working alert from that, so I'm not sure and it'll need to be determined.

> But it won't be triggered when:
[1500, 9000, 9000, 9000]
[1500, 1500, 1500, 9000]

I think I've allowed this to confuse myself. This might not be true. I think in those cases it will actually trigger an alert for the single deviating value, which should be exactly what we want. What Prometheus basically does (explained in Python) would be this:

```python
vs = [1500, 9000, 9000, 9000]
q = np.quantile(vs, 0.5)
fire_alerts = [v for v in vs if v != q]
for value in fire_alerts:
  # fire alert
```
Which is why `[1500, 1500, 9000, 9000]` will result in `len(fire_alerts) == 4`.

That means you're right and we'll only face issues when we have an even amount of values and those two in the middle differ from each other. This is not likely, though I wouldn't say it is impossible. I still expect things to happen which should statistically be impossible, because our initial assumptions and predictions of the future might not always be true.

I'd prefer a solution that'd always work but I won't stop this solution from being merged, neither do I know if such a better solution exist.

Reply:
@p-se perfect then! So, in any case the `topk` will at least return an existing value (not an interpolation), so regardless it might randomly pick 1500 or 9000 on an even distribution (50% each), we don't really care about that scenario as it's a transient one. I guess that's an improvement over the 0.5-quantile approach.

I'm still concerned @aaSharma14 , about the fact that different interfaces might intentionally have different MTUs (usually you only want jumbo frames for the NICs that are going to receive/xmit lots of data, but you will probably leave  1500 byte MTU for provisioning/management interfaces, as jumboframes bring complexity to network config as well). Can we have this scenario covered: comparing a `node_network_mtu_bytes{interface=<nic>}` with the mode (`topk(1, node_network_mtu_bytes{interface=<nic>})`) of those belonging to the same `interface`?